### PR TITLE
Ipfs cid v1 base32

### LINKED ISSUE #5724: Security concern - IPFS ENS Resolving
**Describe the issue**

Using ```gateway.ipfs.io``` when resolving ```ENS``` sites is dangerous. It allows all ```dapps``` to have access over eachothers cookies as well as over the local storage. Research should be done into how this could be mitigated in order to protect users. 

Depending on how the new metamask opt-in feature works, this may also be dangerous. If it is domain wide, all dapps called through their ```ENS``` name are automatically permitted to use metamask.

Could this be done by potentially adding a custom host to a users ```hosts``` file that resolves all ```*.ens``` domains to a custom local server which then simply replaces content?

**Potentially helpful examples**
 - https://github.com/cpacia/ens-chrome-extension
 - https://chrome.google.com/webstore/detail/peername/kkdihlopcnkjinfjhbeopjfmnfpcoaop?hl=en

### PR DESCRIPTION
**Pull request by pldespaigne**
Fixes #5724 

> Recent IPFS update let user query gateways with the following url construction `<ipfs-hash>.gateway.com`. This url construct break same origin policy. Thus making ipfs resolved web site more secured as they will not share cookies & locaStorage anymore.

> The problem is that usual ipfs hash are base58 encoded and contains capital letters, and so it cannot be used in the host part of urls. Before being resolved ipfs hash must be converted to a base32 encoding. To do that I have added a [simple helper function](https://github.com/pldespaigne/content-hash/blob/27dfe53ffc85a44467d466b2d70f5d9f5645efd4/src/helpers.js#L26) to my content-hash library ([already used](https://github.com/MetaMask/metamask-extension/pull/6402) to resolve ENS content).

#### This PR add the following features to MetaMask :
- [x] use new content-hash helper function to convert base58 ipfs hash into base32 (url compatible)
- [x] use new orgin secure ipfs gateway by default
- [x] let the user choose the ipfs gateway in "settings > advanced" menu

#### To test it you can :
1) load this PR's MetaMask version in your browser
2) use the Goerli network
3) go to https://wikipedia.secure-origin.eth
4) open browser developer console and set a random value into the localStorage
5) go to https://other.secure-origin.eth
6) re open localStorage, you should not see the previous value
7) *(optional) repeat without this PR updated code and you should be able to see the same localStorage on every .eth website*

### CODE REVIEW

**Comment by pldespaigne** [comment]
@rekmarks 🙌

**Comment by rekmarks** [comment]
Awesome! Here's the unit test output from CircleCI:
```bash
1312 passing (10s)
  2 failing

  1) AdvancedTab Component
       should render correctly when threeBoxFeatureFlag:

      AssertionError [ERR_ASSERTION]: 10 == 9
      + expected - actual

      -10
      +9
      
      at Context.equal (ui/app/pages/settings/advanced-tab/tests/advanced-tab-component.test.js:19:12)

  2) AdvancedTab Container
       should map state to props correctly:

      AssertionError [ERR_ASSERTION]: { warning: null,
  sendHexData: false,
  advancedInlineGas: false,
  showFiatInTestnets: false,
  autoLogoutTimeLimit: 0,
  t... deepEqual { warning: null,
  sendHexData: false,
  advancedInlineGas: false,
  showFiatInTestnets: false,
  autoLogoutTimeLimit: 0,
  t...
      + expected - actual

       {
         "advancedInlineGas": false
         "autoLogoutTimeLimit": 0
      -  "ipfsGateway": [undefined]
         "sendHexData": false
         "showFiatInTestnets": false
         "threeBoxDisabled": false
         "threeBoxSyncingAllowed": false
      
      at Context.deepEqual (ui/app/pages/settings/advanced-tab/tests/advanced-tab-container.test.js:38:12)
```

**Review by danfinlay — APPROVED** [review]
I've tested this branch and verified its behavior. It redirects to subdomained gateway links, allowing different "decentralized" sites to not collide with cookie storage.

This is an immediate improvement in the security of our ens/IPFS loading, so thank you very much! It's great to see the feature so nicely integrated into our settings page and the rest of our files, this is a very clean contribution.

There are still some improvements we could make, but they can be in a future change:

I believe we may be able to incorporate redirecting so that we could preserve the URL context (either `ens://` or `ipfs://` etc).  This would also allow decentrally-hosted sites to persist cookies even after the gateway is changed.

I think that API may be [webRequest.onBeforeRequest()](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeRequest).

Anyways, just noting that so we can investigate it next.

**Review by danfinlay — DISMISSED** [review]
Oh sorry, before approving, I wanted to check if we'd been in contact with the ipfs gateway that's added here. Are they ready for our traffic?

**Comment by danfinlay** [comment]
It would also be nice if we could resolve `ens://` not just `https://` as I now notice this works with (I've been messing with it. It's fun!)

File: app/scripts/controllers/preferences.js
Comment (pldespaigne) [reply app/scripts/controllers/preferences.js]: I'm not sure about where does this gateway come from, but it was the most cited among various ipfs blog posts. There is also Cloud Flare's gateway at `https://<HASH>.cf-ipfs.com`.

Those are the only two base32 compatible gateways I'm aware of.

Comment (pldespaigne) [reply app/scripts/controllers/preferences.js]: [this](https://github.com/ipfs/ipfs/issues/337#issuecomment-454701521) comment seems to say that `ipfs.dweb.link` is the official gateway

Comment (rekmarks) [reply app/scripts/controllers/preferences.js]: @danfinlay nobody from MetaMask has been in contact with the operators of these gateways. They are listed in repositories and my reading is that they're for public use. I find it hard to imagine that ENS-IPFS sites are so common that we'll overload them any time soon, but we could of course send a courtesy notice.

**Comment by rekmarks** [comment]
> I think that API may be webRequest.onBeforeRequest().

Yeah, that API can definitely be used to accomplish what you're proposing. Just for context, after discussing this with @kumavis, I recommended that @pldespaigne pursue the CIDv1 gateway solution primarily.

File: app/scripts/controllers/preferences.js
Comment (danfinlay) [reply app/scripts/controllers/preferences.js]: > I find it hard to imagine that ENS-IPFS sites are so common that we'll overload them any time soon

I just want to plan for success. The last thing you want to do is irritate your infrastructure just as you're getting traction. Let's reach out before merging.

**Comment by danfinlay** [comment]
> Yeah, that API can definitely be used to accomplish what you're proposing. Just for context, after discussing this with @kumavis, I recommended that @pldespaigne pursue the CIDv1 gateway solution primarily.

Makes sense to me. Just noting that API to make sure we capture it. I've [opened a new issue](https://github.com/MetaMask/metamask-extension/issues/7367) for that extension of this.

Only blocker left here to me is get the go-ahead from ipfs to use that gateway.

**Comment by rekmarks** [comment]
Awaiting answer from IPFS: https://github.com/ipfs/infra/issues/493

**Comment by rekmarks** [comment]
@pldespaigne IPFS recommends that we switch to the dedicated ENS Gateway `eth.link`, see issue comment here: https://github.com/ipfs/infra/issues/493#issuecomment-554346551

Can you make that switch and add the IPFS gateway currently used as a fallback? The bounty deadline has been extended.

**Comment by danfinlay** [comment]
Ehh, hold on to that thought an extra second. I don't think they were accounting for the fact that some of our users run full ethereum nodes, and would also configure IPFS nodes, so this eth-link solution is actually a hard turn for the unconfigurable, and a fair amount of work more.

Let's take a little more time, maybe work with infura, to get an ipfs gateway that we can use as default, and possibly continue using this approach for now.

**Comment by pldespaigne** [comment]
just updated `content-hash` with @lidel 's PR and published it as 2.5.2

**Comment by danfinlay** [comment]
A little update, we have not found anyone who wants to freely host IPFS content in this format yet. We may need to explore an option where we use `js-ipfs` by default, and allow users to configure their own ipfs node if they have access to one.

File: app/scripts/lib/ens-ipfs/setup.js
Comment (whymarrh) [inline app/scripts/lib/ens-ipfs/setup.js]: Does this format work with the old gateway? That is, can someone still use `gateway.ipfs.io`?

File: ui/app/store/actions.js
Comment (whymarrh) [inline ui/app/store/actions.js]: Should this be inside the callback above?

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (whymarrh) [inline ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: Instead of throwing here, can we set the `ipfsGatewayError` state?

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (whymarrh) [inline ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: How does this handle a gateway a user enters that includes the `https://` prefix?

**Review by whymarrh — CHANGES_REQUESTED** [review]

File: app/scripts/lib/ens-ipfs/setup.js
Comment (rekmarks) [reply app/scripts/lib/ens-ipfs/setup.js]: I don't think so. In my understanding the motivating issue encourages us to forbid using that gateway: #5724 

Perhaps we need to explicitly disallow it?

Comment (whymarrh) [reply app/scripts/lib/ens-ipfs/setup.js]: > Perhaps we need to explicitly disallow it?

We could, but maybe that's a separate issue, if we're not trying to support it then this makes sense.

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (rekmarks) [reply ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: It will get caught by the `try` and do so in the `catch` block.

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (rekmarks) [reply ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: It does not, pushing up a fix.

File: app/scripts/lib/ens-ipfs/setup.js
Comment (rekmarks) [reply app/scripts/lib/ens-ipfs/setup.js]: That's what I did, fix incoming.

File: ui/app/store/actions.js
Comment (rekmarks) [reply ui/app/store/actions.js]: I put it inside the callback, only calling it if there's no error.

File: ui/app/pages/settings/advanced-tab/advanced-tab.component.js
Comment (rekmarks) [reply ui/app/pages/settings/advanced-tab/advanced-tab.component.js]: The `https` was only appended for the purpose of checking if it was a valid URL. This is now done invisibly in the change and save handlers.

File: app/_locales/en/messages.json
Comment (whymarrh) [inline app/_locales/en/messages.json]:
```suggestion
    "message": "Enter the URL of the IPFS gateway to use for ENS content resolution"
```

**Review by whymarrh — DISMISSED** [review]

**Comment by whymarrh** [comment]
@rekmarks I've added one last comment, then this LGTM

File: app/_locales/en/messages.json
Comment (rekmarks) [reply app/_locales/en/messages.json]: This was fixed!

**Review by rekmarks — APPROVED** [review]
Approving after review by @whymarrh and earlier review by @danfinlay

**Review by whymarrh — APPROVED** [review]
LGTM

File: app/scripts/lib/ens-ipfs/setup.js
Comment (lidel) [reply app/scripts/lib/ens-ipfs/setup.js]: (sorry for being late, this is probably something for a separate issue/PR now, feel free to copy/paste in correct place) 

To make this more bullet proof, you could add a quick validation, to ensure provided hostname is a real [subdomain gateway](https://docs.ipfs.io/guides/guides/addressing/#subdomain-gateway), and refuse to save invalid ones.

Try fetching a small, well-known hash (for example: string `hello` is `bafkreicysg23kiwv34eg2d7qweipxwosdo2py4ldv42nbauguluen5v6am`) and see if returned value matches expectations:

- https://bafkreicysg23kiwv34eg2d7qweipxwosdo2py4ldv42nbauguluen5v6am.ipfs.dweb.link
- https://bafkreicysg23kiwv34eg2d7qweipxwosdo2py4ldv42nbauguluen5v6am.ipfs.cf-ipfs.com

This would also protect you from users setting [path-based gateways](https://docs.ipfs.io/guides/guides/addressing/#path-gateway) (which do not provide Origin-based isolation.

Comment (rekmarks) [reply app/scripts/lib/ens-ipfs/setup.js]: @lidel thanks for the suggestion, and no worries about missing the merge. New issue here: #7698

**Pull request merged**
